cupy/_core/numpy_allocator.pyx

```diff
@@ -0,0 +1,77 @@
+# distutils: language = c++
+
+from libc cimport errno
+from libc cimport stdlib
+from libc.stdint cimport intptr_t
+from libc.string cimport memset, memcpy
+
+
+IF UNAME_SYSNAME == "Windows":
+    cdef extern from "stdlib.h" nogil:
+        void * _aligned_malloc(size_t size, size_t alignment)
+        void _aligned_free(void * memblock)
+
+    cdef inline void * aligned_alloc(size_t alignment, size_t size) nogil:
+        return _aligned_malloc(size, alignment)
+
+    cdef inline void aligned_free(void * ptr) nogil:
+        _aligned_free(ptr)
+ELSE:
+    cdef extern from * nogil:
+        void * aligned_alloc(size_t alignment, size_t size)
+
+    cdef inline void aligned_free(void * ptr) nogil:
+        stdlib.free(ptr)
+
+
+# CuPy mempool requirement, see ALLOCATION_UNIT_SIZE in cupy/cuda/memory.pyx
+DEF ALIGNMENT = 512
+
+
+cdef extern from *:
+    """
+    extern "C" {
+        void* _calloc(size_t nmemb, size_t size);
+        void* _malloc(size_t size);
+        void* _realloc(void *ptr, size_t size);
+        void _free(void* ptr);
+    }
+    """
+    void* _calloc(size_t nmemb, size_t size) nogil
+    void* _malloc(size_t size) nogil
+    void* _realloc(void *ptr, size_t size) nogil
+    void _free(void* ptr) nogil
```

Comment:
I think I was probably wrong in suggesting this in a past inline comment. We should just set `CYTHON_EXTERN_C` at build time, see e.g.
- https://cython.readthedocs.io/en/latest/src/userguide/external_C_code.html#c-public-declarations
- https://github.com/rapidsai/rmm/pull/1350

Reply:
fixed in [c9298fb](https://github.com/cupy/cupy/pull/8442/commits/c9298fba9500b3a0c619b13596f90bdc25cd8fdc)